tests/jdom/fastforwardtester.ts

```diff
@@ -0,0 +1,187 @@
+        if (deviceIds.length === 0) {
+            return Promise.resolve()
+        }
+
+        return new Promise(resolve => {
+            const devicesIdSet = new Set(deviceIds)
+            const announcedIdSet = new Set()
+            const onHandler = (device: JDDevice) => {
+                if (devicesIdSet.has(device.deviceId)) {
+                    announcedIdSet.add(device.deviceId)
+                }
+                if (setEquals(devicesIdSet, announcedIdSet)) {
+                    this.bus.off(DEVICE_ANNOUNCE, onHandler)
+                    resolve(undefined)
+                }
+            }
+            this.bus.on(DEVICE_ANNOUNCE, onHandler)
+        })
+    }
+
+    // TODO is there a better abstraction where we can create the service as independent function calls
+    // but only have one delay? Would need some way to overlap the creation, and
+    // separate the promise / await portion.
+    // Perhaps have createService and fast-fowrad await?
+    public async createServices<T extends Record<string, JDServiceServer>>(
+        servers: T
+    ): Promise<{ [key in keyof T]: CreatedServerService<T[key]> }> {
+        // attach servers to the bus as devices
+        const devices = Object.entries(servers).map(([name, server]) => {
+            const device = this.bus.addServiceProvider(
+                new JDServiceProvider([server])
+            )
+            return {
+                name: name,
+                server: server,
+                device: device,
+            }
+        })
+
+        // Wait for created devices to be announced, so services become available
+        const deviceIds = devices.map(elt => elt.device.deviceId)
+        await (this.bus.scheduler as FastForwardScheduler).runToPromise(
+            this.waitForAnnounce(deviceIds)
+        )
+
+        // Create the output map
+        const namesToServices: [
+            string,
+            CreatedServerService<JDServiceServer>
+        ][] = devices.map(({ name, server, device }) => {
+            const services = device.services({
+                serviceClass: server.serviceClass,
+            })
+            assert(
+                services.length > 0,
+                `created device ${device.friendlyName} has no service of ${server.specification.name}`
+            )
+            assert(
+                services.length == 1,
+                `created device ${device.friendlyName} has multiple service of ${server.specification.name}`
+            )
+            return [
+                name,
+                {
+                    server: server,
+                    service: services[0],
+                },
+            ]
+        })
+
+        const namesToServicesObject = namesToServices.reduce(
+            (last, [name, service]) => Object.assign(last, { [name]: service }),
+            {}
+        )
+
+        return namesToServicesObject as {
+            [key in keyof T]: CreatedServerService<T[key]>
```

Comment:
maybe you can preallocate the service tester by returning
```
 {
    services: { [key in keyof T]: CreatedServerService<T[key]> }
    testers: { [key in keyof T]: ServiceTester<T[key]> }
}
```
not sure if it helps.

Reply:
I don't think that's necessary at this point. I expect more tests to use virtual servers as test inputs, instead of having assertions on them.